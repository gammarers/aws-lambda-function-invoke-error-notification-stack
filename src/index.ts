import * as crypto from 'crypto';
import * as cdk from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';

export interface NotificationsProperty {
  readonly emails?: string[];
}
export interface LambdaFunctionInvokeErrorNotificationStackProps extends cdk.StackProps {
  readonly notifications: NotificationsProperty;
}

export class LambdaFunctionInvokeErrorNotificationStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: LambdaFunctionInvokeErrorNotificationStackProps) {
    super(scope, id, props);

    const random = crypto.createHash('shake256', { outputLength: 4 })
      .update(cdk.Names.uniqueId(this))
      .digest('hex');

    // SNS„Éà„Éî„ÉÉ„ÇØ„ÅÆ‰ΩúÊàê
    const topic: sns.Topic = new sns.Topic(this, 'LambdaFunctionInvokeErrorNotificationTopic', {
      topicName: `lambda-funcs-invoke-error-notification-${random}-topic`,
      displayName: 'Lambda Functions Invoke Error Notification Topic',
    });

    // Subscribe an email endpoint to the topic
    for (const email of props.notifications.emails ?? []) {
      topic.addSubscription(new subscriptions.EmailSubscription(email));
    }

    // Prepare Message
    const prepareMessage: sfn.Pass = new sfn.Pass(this, 'PrepareMessage', {
      parameters: {
        Subject: sfn.JsonPath.format('üòµ [Failur] AWS Lambda Function Invocation Failur Notification [{}][{}]',
          sfn.JsonPath.stringAt('$.account'),
          sfn.JsonPath.stringAt('$.region'),
        ),
        Message: sfn.JsonPath.format('Account : {}\nRegion : {}\nFunction : {}\nErrorMessage : {}\nTrace : \n{}',
          sfn.JsonPath.stringAt('$.account'),
          sfn.JsonPath.stringAt('$.region'),
          sfn.JsonPath.stringAt('$.detail.requestContext.functionArn'),
          sfn.JsonPath.stringAt('$.detail.responsePayload.errorMessage'),
          sfn.JsonPath.stringAt('$.Prepare.Concatenated.Trace'),
        ),
      },
      resultPath: '$.Prepare.Sns.Topic',
    });

    const init: sfn.Pass = new sfn.Pass(this, 'Init', {
      result: sfn.Result.fromString(''),
      resultPath: '$.Prepare.Concatenated.Trace',
    });

    const traceLinces: sfn.Pass = new sfn.Pass(this, 'TraceLinces', {
      parameters: {
        Lines: sfn.JsonPath.stringAt('$.detail.responsePayload.trace'),
      },
      resultPath: '$.TempTrace',
    });

    init.next(traceLinces);

    const getTraceLine = new sfn.Pass(this, 'GetTraceLine', {
      parameters: {
        Line: sfn.JsonPath.arrayGetItem(sfn.JsonPath.stringAt('$.TempTrace.Lines'), 0),
      },
      resultPath: '$.Temp.GetTrace',
    });

    const checkUntreatedTranceLinesExist: sfn.Choice = new sfn.Choice(this, 'CheckUntreatedTranceLinesExist')
      .when(sfn.Condition.isPresent('$.TempTrace.Lines[0]'), getTraceLine)
      .otherwise(prepareMessage);

    traceLinces.next(checkUntreatedTranceLinesExist);

    const concatenateValue: sfn.Pass = new sfn.Pass(this, 'Concatenate', {
      parameters: {
        Trace: sfn.JsonPath.format('{}{}\n', sfn.JsonPath.stringAt('$.Prepare.Concatenated.Trace'), sfn.JsonPath.stringAt('$.Temp.GetTrace.Line')),
      },
      resultPath: '$.Prepare.Concatenated',
    });

    getTraceLine.next(concatenateValue);

    const untreatedTranceLines: sfn.Pass = new sfn.Pass(this, 'UntreatedTranceLines', {
      parameters: {
        Lines: sfn.JsonPath.stringAt('$.TempTrace.Lines[1:]'),
      },
      resultPath: '$.TempTrace',
    });

    concatenateValue.next(untreatedTranceLines);

    untreatedTranceLines.next(checkUntreatedTranceLinesExist);

    const sendNotification: tasks.SnsPublish = new tasks.SnsPublish(this, 'SendNotification', {
      topic: topic,
      inputPath: '$.Prepare.Sns.Topic',
      subject: sfn.JsonPath.stringAt('$.Subject'),
      message: sfn.TaskInput.fromJsonPathAt('$.Message'),
      resultPath: '$.Result.Sns.Topic',
    });

    prepareMessage.next(sendNotification);

    // Step Functions State Machine
    const stateMachine: sfn.StateMachine = new sfn.StateMachine(this, 'StateMachine', {
      stateMachineName: `lambda-func-invoke-error-notification-${random}-state-machine`,
      timeout: cdk.Duration.minutes(5),
      definitionBody: sfn.DefinitionBody.fromChainable(init),
    });

    // EventBridge Rule
    new events.Rule(this, 'LambdaFunctionInvokeErrorCatchRule', {
      ruleName: `lambda-func-invoke-error-catch-${random}-rule`,
      eventPattern: {
        source: ['lambda'],
        detailType: ['Lambda Function Invocation Result - Failure'],
        detail: {
          responsePayload: {
            errorType: [
              'error',
            ],
          },
        },
      },
      targets: [
        new targets.SfnStateMachine(stateMachine, {
          role: new iam.Role(this, 'StartExecMachineRole', {
            roleName: `lambda-func-invoke-error-start-exec-machine-${random}-role`,
            description: 'lambda func invoke error start exec machine (send notification).',
            assumedBy: new iam.ServicePrincipal('events.amazonaws.com'),
            inlinePolicies: {
              'states-start-execution-policy': new iam.PolicyDocument({
                statements: [
                  new iam.PolicyStatement({
                    effect: iam.Effect.ALLOW,
                    actions: [
                      'states:StartExecution',
                    ],
                    resources: ['*'],
                  }),
                ],
              }),
            },
          }),
        }),
      ],
    });
  }
}