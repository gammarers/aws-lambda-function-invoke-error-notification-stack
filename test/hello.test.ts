import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LambdaFunctionInvokeErrorNotificationStack } from '../src';

test('hello', () => {
  const app = new App();

  const stack = new LambdaFunctionInvokeErrorNotificationStack(app, 'LambdaFunctionInvokeErrorNotificationStack', {
    notifications: {
      emails: [
        'foo@example.com',
      ],
    },
  });

  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});