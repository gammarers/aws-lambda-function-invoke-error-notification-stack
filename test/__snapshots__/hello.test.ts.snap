// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LambdaFunctionInvokeErrorCatchRule70C95FF4": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "responsePayload": {
              "errorType": [
                "error",
              ],
            },
          },
          "detail-type": [
            "Lambda Function Invocation Result - Failure",
          ],
          "source": [
            "lambda",
          ],
        },
        "Name": "lambda-func-invoke-error-catch-38c73628-rule",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "StateMachine2E01A3A5",
            },
            "Id": "Target0",
            "RoleArn": {
              "Fn::GetAtt": [
                "StartExecMachineRole28D483E3",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "LambdaFunctionInvokeErrorNotificationTopic54C7E386": {
      "Properties": {
        "DisplayName": "Lambda Functions Invoke Error Notification Topic",
        "TopicName": "lambda-funcs-invoke-error-notification-38c73628-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "LambdaFunctionInvokeErrorNotificationTopicfooexamplecom3EFA0BE4": {
      "Properties": {
        "Endpoint": "foo@example.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "LambdaFunctionInvokeErrorNotificationTopic54C7E386",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "StartExecMachineRole28D483E3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "lambda func invoke error start exec machine (send notification).",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "states:StartExecution",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "states-start-execution-policy",
          },
        ],
        "RoleName": "lambda-func-invoke-error-start-exec-machine-38c73628-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "StartExecMachineRoleDefaultPolicy9C23DAB3": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "StateMachine2E01A3A5",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StartExecMachineRoleDefaultPolicy9C23DAB3",
        "Roles": [
          {
            "Ref": "StartExecMachineRole28D483E3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"PrepareMessage","States":{"PrepareMessage":{"Type":"Pass","ResultPath":"$.Prepare.Sns.Topic","Parameters":{"Subject.$":"States.Format('ðŸ˜µ [Failur] AWS Lambda Function Invocation Failur Notification [{}][{}]', $.account, $.region)","Message.$":"States.Format('Account : {}\\nRegion : {}\\nFunction : {}\\nErrorMessage : {}\\nTrace : {}', $.account, $.region, $.detail.requestContext.functionArn, $.detail.responsePayload.errorMessage, $.detail.responsePayload.trace)"},"Next":"SendNotification"},"SendNotification":{"End":true,"Type":"Task","InputPath":"$.Prepare.Sns.Topic","ResultPath":"$.Result.Sns.Topic","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish","Parameters":{"TopicArn":"",
              {
                "Ref": "LambdaFunctionInvokeErrorNotificationTopic54C7E386",
              },
              "","Message.$":"$.Message","Subject.$":"$.Subject"}}},"TimeoutSeconds":300}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
        "StateMachineName": "lambda-func-invoke-error-notification-38c73628-state-machine",
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineRoleB840431D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "LambdaFunctionInvokeErrorNotificationTopic54C7E386",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
